src/dragonball/src/dbs_pci/src/msi.rs

```diff
@@ -0,0 +1,759 @@
+                x if addr_hi_offset.is_some() && x == addr_hi_offset.unwrap() => {
+                    LittleEndian::write_u32(data, self.msg_addr_hi);
+                }
+                x if x == msg_data_offset => {
+                    LittleEndian::write_u32(data, self.msg_data as u32);
+                }
+                x if mask_bits_offset.is_some() && x == mask_bits_offset.unwrap() => {
+                    LittleEndian::write_u32(data, self.mask_bits);
+                }
+                x if mask_bits_offset.is_some() && x == mask_bits_offset.unwrap() + 4 => {
+                    self.get_pending_state(data, 0);
+                }
+                _ => debug!("invalid offset"),
+            },
+            _ => debug!("invalid data length"),
+        }
+
+        Ok(())
+    }
+
+    /// Handle write accesses to the PCI MSI capability structure.
+    pub fn write(&mut self, offset: u64, data: &[u8]) -> std::io::Result<()> {
+        let (msg_data_offset, addr_hi_offset, mask_bits_offset) = self.get_offset_info();
+
+        // Update cache without overriding the read-only bits.
+        match data.len() {
+            1 => match offset {
+                x if mask_bits_offset.is_some() && x == mask_bits_offset.unwrap() => {
+                    self.mask_bits = data[0] as u32 | (self.mask_bits & 0xffff_ff00);
+                }
+                x if mask_bits_offset.is_some() && x == mask_bits_offset.unwrap() + 1 => {
+                    self.mask_bits = (data[0] as u32) << 8 | (self.mask_bits & 0xffff_00ff);
+                }
+                x if mask_bits_offset.is_some() && x == mask_bits_offset.unwrap() + 2 => {
+                    self.mask_bits = (data[0] as u32) << 16 | (self.mask_bits & 0xff00_ffff);
+                }
+                x if mask_bits_offset.is_some() && x == mask_bits_offset.unwrap() + 3 => {
+                    self.mask_bits = (data[0] as u32) << 24 | (self.mask_bits & 0x00ff_ffff);
+                }
+                _ => debug!("invalid offset"),
+            },
+            2 => {
+                let value = LittleEndian::read_u16(data);
+                match offset {
+                    MSI_MSG_CTL_OFFSET => {
+                        self.msg_ctl = (self.msg_ctl
+                            & !(MSI_CTL_ENABLE | MSI_CTL_MULTI_MSG_ENABLE))
+                            | value & (MSI_CTL_ENABLE | MSI_CTL_MULTI_MSG_ENABLE);
+                    }
+                    x if x == msg_data_offset => {
+                        self.msg_data = value;
+                    }
+                    x if mask_bits_offset.is_some() && x == mask_bits_offset.unwrap() => {
+                        self.mask_bits = value as u32 | (self.mask_bits & 0xffff_0000);
+                    }
+                    x if mask_bits_offset.is_some() && x == mask_bits_offset.unwrap() + 2 => {
+                        self.mask_bits = (value as u32) << 16 | (self.mask_bits & 0x0000_ffff);
+                    }
```

Comment:
Ditto.

Reply:
this is for writing bits according to pci spec. I have added a summarised comment above
"Support writing mask_bits, MSI control information, message data, message address."